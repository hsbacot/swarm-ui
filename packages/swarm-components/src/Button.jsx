// @flow
import * as React from 'react';
import Icon from './Icon';
import {
	getButtonType,
	getSwarmSize,
	getIconPosition,
	getButtonWidth,
} from './utils/buttonUtils';

export type ButtonProps = {
	/**
	 * The bordered button
	 */
	bordered?: boolean,

	/**
	 * Classes for additional styles to be applied
	 */
	className?: string,

	/**
	 * Indicates whether the button is disabled or not
	 */
	disabled?: boolean,

	/**
	 * Supports icon shapes found at https://swarm-docs.now.sh/Foundations/Icons
	 */
	icon?: string,

	/**
	 * Strips all styles from the button
	 */
	reset?: boolean,

	/**
	 * Inverted styles for dark backgrounds
	 */
	inverted?: boolean,

	/**
	 * The neutral style
	 */
	neutral?: boolean,

	/**
	 * The normal width style (3x padding)
	 */
	normal?: boolean,

	/**
	 * The function invoked when interacting with Button
	 */
	onClick?: () => {},

	/**
	 * The primary style
	 */
	primary?: boolean,

	/**
	 * Aligns the icon to the right
	 */
	right?: boolean,

	/**
	 * The large button size
	 */

	large?: boolean,

	/**
	 * Use the small button size
	 */
	small?: boolean,

	/**
	 * Grow button to 100% width (full width)
	 */
	grow?: boolean,

	/**
	 *
	 */
	forwardedRef?: ?React.ElementRef<*>,

	/**
	 * Content of button
	 */
	children: React.Node,
};

class Button extends React.Component<ButtonProps> {
	// destructuring to not pass invalid attributes to node
	render() {
		const {
			icon,
			right,
			grow, // eslint-disable-line no-unused-vars
			children,
			bordered, // eslint-disable-line no-unused-vars
			neutral, // eslint-disable-line no-unused-vars
			normal, // eslint-disable-line no-unused-vars
			primary, // eslint-disable-line no-unused-vars
			inverted, // eslint-disable-line no-unused-vars
			reset, // eslint-disable-line no-unused-vars
			small, // eslint-disable-line no-unused-vars
			large, // eslint-disable-line no-unused-vars
			forwardedRef,
			...other
		} = this.props;

		const buttonType = getButtonType(this.props);

		return (
			<button
				data-swarm-button={buttonType}
				data-swarm-size={getSwarmSize(this.props)}
				data-icon={getIconPosition(this.props)}
				data-swarm-width={getButtonWidth(this.props)}
				type="button"
				ref={forwardedRef}
				{...other}
			>
				{icon ? (
					<span>
						{right ? (
							<>
								{children}
								<Icon shape={icon} size="xs" />
							</>
						) : (
							<>
								<Icon shape={icon} size="xs" />
								{children}
							</>
						)}
					</span>
				) : (
					children
				)}
			</button>
		);
	}
}

export const ForwardedButton = React.forwardRef<ButtonProps, HTMLButtonElement>(
	(props, ref) => <Button {...props} forwardedRef={ref} />
);

export default Button;
