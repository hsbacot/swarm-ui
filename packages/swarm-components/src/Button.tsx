// @flow
import * as React from 'react';
import { Icon } from './Icon';
import { getButtonType, getSwarmSize, getIconPosition } from './utils/buttonUtils';

export interface ButtonProps {
	/**
	 * The bordered button
	 */
	bordered?: boolean,

	/**
	 * Classes for additional styles to be applied
	 */
	className?: string,

	/**
	 * Indicates whether the button is disabled or not
	 */
	disabled?: boolean,

	/**
	 * Supports solid icon shapes found at https://swarm-docs.now.sh/Foundations/Icons. This is the recommended way to use icons in buttons.
	 */
	icon?: string,

	/**
	 * Supports icon only padding for using icons as children rather than by passing an icon prop. Using the icon prop is preferred.
	 */
	iconOnly?: boolean,

	/**
	 * Strips all styles from the button
	 */
	reset?: boolean,

	/**
	 * Inverted styles for dark backgrounds
	 */
	inverted?: boolean,

	/**
	 * The neutral style
	 */
	neutral?: boolean,

	/**
	 * The function invoked when interacting with Button
	 */
	onClick?: () => {},

	/**
	 * The primary style
	 */
	primary?: boolean,

	/**
	 * Aligns the icon to the right
	 */
	right?: boolean,

	/**
	 * The size of the button. supports "small", "large", & "default"
	 */
	size?: "small" | "large" | "default",
	/**
	 * The large button size
	 */

	large?: boolean,

	/**
	 * Use the small button size
	 */
	small?: boolean,

	/**
	 * Grow button to 100% width (full width)
	 */
	grow?: boolean,

	/**
	 *
	 */
	forwardedRef?: React.Ref<ButtonProps | null> | any,

	/**
	 * Content of button
	 */
	children?: React.ReactNode,
};

class Button extends React.Component<ButtonProps> {
	// destructuring to not pass invalid attributes to node
	render() {
		const {
			icon,
			iconOnly,
			right,
			grow,
			children,
			bordered, // eslint-disable-line no-unused-vars
			neutral, // eslint-disable-line no-unused-vars
			primary, // eslint-disable-line no-unused-vars
			inverted, // eslint-disable-line no-unused-vars
			reset, // eslint-disable-line no-unused-vars
			size, // eslint-disable-line no-unused-vars
			small, // eslint-disable-line no-unused-vars
			large, // eslint-disable-line no-unused-vars
			forwardedRef,
			...other
		} = this.props;

		const buttonType = getButtonType(this.props);
		const width = grow ? 'grow' : 'default';


		return (
			<button
				data-swarm-button={buttonType}
				data-swarm-size={getSwarmSize(this.props)}
				data-icon={getIconPosition(this.props)}
				data-swarm-width={width}
				type="button"
				ref={forwardedRef}
				{...other}
			>
				{icon ? (
					<span>
						{right ? (
							<>
								{children}
								<Icon shape={icon} />
							</>
						) : (
							<>
								<Icon shape={icon} />
								{children}
							</>
						)}
					</span>
				) : iconOnly ? (
					<span>{children}</span>
				) : (
					children
				)}
			</button>
		);
	}
}

// any for now. Not sure how to type forwarded ref
const ForwardedButton = React.forwardRef<ButtonProps, any>(
	(props, ref) => <Button {...props} forwardedRef={ref} />
);

export { Button, ForwardedButton };
