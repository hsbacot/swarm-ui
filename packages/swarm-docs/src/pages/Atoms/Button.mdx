import { Button } from '@meetup/swarm-components'
import { Icon } from '@meetup/swarm-components'
import Example from '../../components/examples/button';
import PropsTable from '../../components/propstable';
import { Close, Check } from '@meetup/swarm-icons/lib/components/solid';


# Button
---
Buttons are a fundamental component and primarily how users trigger actions. They can be customized by sizes and color to communicate importance and content heirarchy.
<br/><br/>

<Example /><br/>

<br/><br/>


```.jsx
<>
  <Button>Default</Button>
  <Button primary>Primary</Button>
  <Button neutral>Neutral</Button>
  <Button bordered>Bordered</Button>
  <Button icon="send">Send message</Button>
  <Button large iconOnly>
    <Search/>
  </Button>
  <br/><br/>
  <Button neutral grow icon="InfoCircled">Fill the space</Button>
</>
```

<br/><br/>

### Props
<br/>
<PropsTable propMetaData={Button.__docgenInfo} />

<br/><br/>


### Anatomy
All button styles and sizes share the same anatomy.
<br/>
<img src="/docs-images/button-anatomy.png" width="810" height="252" />
<br/>
<div class="grid-2-col">
    <div class="column">
        <p class="text--body"><b>Label</b></p>
        <p class="text--body">Button labels should be short and actionable. Label attributes are driven by the selected button style and size.</p>
    </div>
    <div class="column">
        <p class="text--body"><strong>Border</strong></p>
        <p class="text--body">All buttons have a solid 1px border, though some styles require the same color for background and border, hiding the presence of the border.</p>
    </div>
    <div class="column">
        <p class="text--body"><strong>Background</strong></p>
        <p class="text--body">Button style determines its background color and opacity.</p>
    </div>
    <div class="column">
        <p class="text--body"><strong>Radius</strong></p>
        <p class="text--body">All buttons in all sizes and viewports have an 8px border radius.</p>
    </div>
</div>

<br/><br/>

### Sizing
<br/>

** Size & Height **

Any button can be configured to one of three distinct sizes. Size dictates label text style and top/bottom padding. Button padding uses spacing variables which are dynamic based on viewport.
<br/>
<img src="/docs-images/button-sizing-size.png" width="600" height="405" />

<br/><br/><br/>

** Width **

Buttons will inherit their left and right padding from their size. Use the grow attribute to expand a button to fill its container.
<br/>
<img src="/docs-images/button-sizing-width.png" width="600" height="120" />

<br/><br/><br/><br/>

### a11y
<br/>

** Keyboard interactions**

The button action should be triggered by the enter and space keys.

<br/><br/><br/>

### Content Guidelines
<br/>

** Button labels **

Should clearly indicate what will happen if the user clicks on them. They should be easy to scan and comprehend.

<br/>

<ul style="margin: 0 0 0 20px;">
    <li> Keep it scannable: Avoid articles like "a" and "the"</li>
    <li> Use sentence case</li>
    <li> Don't use first person to speak from the user's point of view (Sign up vs. Sign me up!)</li>
    <li> No ampersands or end punctuation (Agree and continue vs. Agree & continue)</li>
    <li> Make priorities clear: Never present the user with more than one primary action</li>
</ul>


