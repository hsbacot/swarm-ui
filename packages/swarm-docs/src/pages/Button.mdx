import { Button } from '@meetup/swarm-components'
import { Icon } from '@meetup/swarm-components'
import Example from '../components/examples/button';
import PropsTable from '../components/propstable';

# Button
---
Buttons are the primary way to interact with content.

## Examples
The following are all the available states for a `Button`:
<br/>
<Example /><br/>

## Props
All available props for `Button` include:
<br/>
<PropsTable propMetaData={Button.__docgenInfo} />

## Snippet
Try using `Button` with this interactive code snippet. Update a prop and see what happens!

```.jsx
<>
  <Button>Click</Button>
  <br />
  <Button icon="bolt">Click</Button>
</>
```

## Anatomy
All button styles and sizes share the same anatomy.
<img src="/docs-images/button-anatomy.png" width="810" height="252" />
<br/>
<div class="grid-2-col">
    <div class="column">
        <p class="text--body"><strong>Label</strong></p>
        <p class="text--body">Button labels should be short and actionable. Label attributes are driven by button style and size.</p>
    </div>
    <div class="column">
        <p class="text--body"><strong>Border</strong></p>
        <p class="text--body">All buttons have solid 1px borders, though some styles require the same color for background and border, hiding the presence of the border.</p>
    </div>
    <div class="column">
        <p class="text--body"><strong>Background</strong></p>
        <p class="text--body">Button style determines its background color and opacity.</p>
    </div>
    <div class="column">
        <p class="text--body"><strong>Radius</strong></p>
        <p class="text--body">All buttons in all sizes and viewports have an 8px border radius.</p>
    </div>
</div>

## Styles

** Default **

<img src="/docs-images/button-secondary-states.png" width="810" height="132" />
<br/>

** Primary **

<img src="/docs-images/button-primary-states.png" width="810" height="132" />
<br/>

** Neutral **

<img src="/docs-images/button-neutral-states.png" width="810" height="132" />
<br/>

** Bordered **

<img src="/docs-images/button-bordered-states.png" width="810" height="132" />
<br/>

** Disabled **

<img src="/docs-images/button-disabled-states.png" width="250" height="132" />

## Sizing

** Size & Height **

Any button can be configured to one of three distinct sizes. Size dictates label style and top/bottom padding. Button padding uses spacing variables which are dynamic based on viewport.
<br/>
<img src="/docs-images/button-sizing-size.png" width="600" height="373" />
<br/><br/>

** Width **

Any button can be configured to one of three width settings. Width dictates left/right padding behavior. Button padding uses spacing variables which are dynamic based on viewport.
<br/>
<img src="/docs-images/button-sizing-width.png" width="600" height="371" />
<br/><br/>

## a11y
Keyboard interactions: The button action should be triggered by the enter and space keys.
