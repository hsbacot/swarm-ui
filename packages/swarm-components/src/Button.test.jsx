import React from 'react';
import { shallow } from 'enzyme';
import { Button } from './Button';

describe('Button', () => {
	const testCases = [
		['Default', <Button key="1">Press me</Button>],
		[
			'Disabled',
			<Button disabled key="2">
				Can&apos;t press me
			</Button>,
		],
		[
			'Primary',
			<Button primary key="3">
				Must press me
			</Button>,
		],
		[
			'Bordered',
			<Button bordered key="4">
				Press me
			</Button>,
		],
		[
			'Icon with text',
			<Button icon="bolt" key="5">
				Press me
			</Button>,
		],
		[
			'Icon (right) with text',
			<Button right icon="chevron-right" key="6">
				Press me
			</Button>,
		],
		['Icon without text', <Button icon="bolt" key="7" />],
		[
			'Reset',
			<Button reset key="8">
				Press me
			</Button>,
		],
		[
			'Inverted',
			<Button inverted key="9">
				Press me
			</Button>,
		],
		[
			'Neutral',
			<Button neutral key="10">
				Press me
			</Button>,
		],
		[
			'Small',
			<Button small key="11">
				Press me
			</Button>,
		],
		[
			'Grow',
			<Button grow key="12">
				Full Width Button
			</Button>,
		],
	];

	test.each(testCases)('Snapshot: %s', (description, element) => {
		expect(shallow(element)).toMatchSnapshot();
	});
});
