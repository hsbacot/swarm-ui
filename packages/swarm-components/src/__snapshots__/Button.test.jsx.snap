// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Snapshot: Bordered 1`] = `
<button
  data-icon="left"
  data-swarm-button="bordered"
  data-swarm-size="default"
  data-swarm-width="default"
  type="button"
>
  Press me
</button>
`;

exports[`Button Snapshot: Default 1`] = `
<button
  data-icon="left"
  data-swarm-button="default"
  data-swarm-size="default"
  data-swarm-width="default"
  type="button"
>
  Press me
</button>
`;

exports[`Button Snapshot: Disabled 1`] = `
<button
  data-icon="left"
  data-swarm-button="disabled"
  data-swarm-size="default"
  data-swarm-width="default"
  disabled={true}
  type="button"
>
  Can't press me
</button>
`;

exports[`Button Snapshot: Grow 1`] = `
<button
  data-icon="left"
  data-swarm-button="default"
  data-swarm-size="default"
  data-swarm-width="grow"
  type="button"
>
  Full Width Button
</button>
`;

exports[`Button Snapshot: Icon (right) with text 1`] = `
<button
  data-icon="right"
  data-swarm-button="default"
  data-swarm-size="default"
  data-swarm-width="default"
  type="button"
>
  <span>
    Press me
    <Icon
      shape="chevron-right"
    />
  </span>
</button>
`;

exports[`Button Snapshot: Icon with text 1`] = `
<button
  data-icon="left"
  data-swarm-button="default"
  data-swarm-size="default"
  data-swarm-width="default"
  type="button"
>
  <span>
    <Icon
      shape="bolt"
    />
    Press me
  </span>
</button>
`;

exports[`Button Snapshot: Icon without text 1`] = `
<button
  data-icon="only"
  data-swarm-button="default"
  data-swarm-size="default"
  data-swarm-width="default"
  type="button"
>
  <span>
    <Icon
      shape="bolt"
    />
  </span>
</button>
`;

exports[`Button Snapshot: Inverted 1`] = `
<button
  data-icon="left"
  data-swarm-button="inverted"
  data-swarm-size="default"
  data-swarm-width="default"
  type="button"
>
  Press me
</button>
`;

exports[`Button Snapshot: Neutral 1`] = `
<button
  data-icon="left"
  data-swarm-button="neutral"
  data-swarm-size="default"
  data-swarm-width="default"
  type="button"
>
  Press me
</button>
`;

exports[`Button Snapshot: Primary 1`] = `
<button
  data-icon="left"
  data-swarm-button="primary"
  data-swarm-size="default"
  data-swarm-width="default"
  type="button"
>
  Must press me
</button>
`;

exports[`Button Snapshot: Reset 1`] = `
<button
  data-icon="left"
  data-swarm-button="reset"
  data-swarm-size="default"
  data-swarm-width="default"
  type="button"
>
  Press me
</button>
`;

exports[`Button Snapshot: Small 1`] = `
<button
  data-icon="left"
  data-swarm-button="default"
  data-swarm-size="small"
  data-swarm-width="default"
  type="button"
>
  Press me
</button>
`;
